projects/RabbitMQ.Client/client/impl/RabbitMQActivitySource.cs

```diff
@@ -0,0 +1,167 @@
+                }
+
+                return activity;
+            }
+
+            return null;
+        }
+
+        internal static Activity Process(BasicDeliverEventArgs deliverEventArgs)
+        {
+            if (source.HasListeners())
+            {
+                // Extract the PropagationContext of the upstream parent from the message headers.
+                DistributedContextPropagator.Current.ExtractTraceIdAndState(deliverEventArgs.BasicProperties, ExtractTraceIdAndState, out string traceId, out string traceState);
+                IEnumerable<KeyValuePair<string, string>> baggage = DistributedContextPropagator.Current.ExtractBaggage(deliverEventArgs.BasicProperties, ExtractTraceIdAndState);
+                ActivityContext.TryParse(traceId, traceState, out ActivityContext parentContext);
+                Activity activity = StartRabbitMQActivity($"{deliverEventArgs.RoutingKey} process", ActivityKind.Consumer, parentContext);
+                if (activity != null && activity.IsAllDataRequested)
+                {
+                    if (baggage != null)
+                    {
+                        foreach (var item in baggage)
+                        {
+                            Activity.Current?.SetBaggage(item.Key, item.Value);
+                        }
+                    }
+                    
+                    PopulateMessagingTags("process", deliverEventArgs.RoutingKey, deliverEventArgs.Exchange, deliverEventArgs.DeliveryTag, deliverEventArgs.BasicProperties, deliverEventArgs.Body.Length, activity);
+                }
+
+                return activity;
+            }
+
+            return null;
+        }
+
+        private static Activity StartRabbitMQActivity(string name, ActivityKind kind, ActivityContext parentContext = default)
+        {
+            Activity activity = source.StartActivity(name, kind, parentContext);
+            if (activity != null && activity.IsAllDataRequested)
+            {
+                if (!string.IsNullOrEmpty(Thread.CurrentThread.Name))
+                {
+                    activity.SetTag("thread.name", Thread.CurrentThread.Name);
+                }
+
+                activity
+                    .SetTag("thread.id", Thread.CurrentThread.ManagedThreadId.ToString())
+                    .SetTag("net.host.name", Environment.MachineName)
+                    .SetTag(MessagingSystem, "rabbitmq")
+                    .SetTag(ProtocolName, "amqp")
+                    .SetTag(ProtocolVersion, "0.9.1");
+            }
+
+            return activity;
+        }
+
+        private static void PopulateMessagingTags(string operation, string routingKey, string exchange, ulong deliveryTag, IReadOnlyBasicProperties readOnlyBasicProperties, int bodySize, Activity activity)
+        {
+            
+            activity
+            .SetTag(MessagingOperation, operation == "publish" ? operation : null)
```

Comment:
Why only for `publish`?
`receive` and `process` are also allowed.

I would put it as const strings, to suggest that values should be treated as enum.

Reply:
Right, forgot those. They weren't needed before, but I'll add them now :)